src/Build/BackEnd/BuildManager/BuildManager.cs

```diff
@@ -2184,20 +2184,13 @@ private void OnThreadException(Exception e)
                             continue;
                         }
 
-                        submission.CompleteLogging(false);
-
                         // Attach the exception to this submission if it does not already have an exception associated with it
-                        if (submission.BuildResult?.Exception == null)
+                        if (!submission.IsCompleted && submission.BuildResult != null && submission.BuildResult.Exception == null)
                         {
-                            if (submission.BuildResult == null)
-                            {
-                                submission.BuildResult = new BuildResult(submission.BuildRequest, e);
-                            }
-                            else
-                            {
-                                submission.BuildResult.Exception = _threadException.SourceException;
-                            }
+                            submission.BuildResult.Exception = e;
                         }
+                        submission.CompleteLogging(false);
+                        submission.CompleteResults(new BuildResult(submission.BuildRequest, e));
```

Comment:
`CompleteResults` doesn't change the result if already non-null so this matches the old behavior.